addons/l10n_ro_efactura/models/account_move.py

```diff
@@ -0,0 +1,209 @@
+
+    def _l10n_ro_edi_fetch_invoice_sending_documents(self):
+        """ This method loops over all invoice with sending document in `self`. For each of them,
+            it pre-checks error and make a fetch request for the invoice. Based on the answer, it will then:
+             - if error -> delete all errors, create a new error document
+             - if receive `key_download` -> immediately make a download request and process it:
+                - if error -> delete all sending and error documents, create a new error document
+                - if success -> delete all sending and error documents, create success document
+             - if no answer is received, it will do nothing on the selected invoice
+        """
+        session = requests.session()
+        to_delete_documents = self.env['l10n_ro_edi.document']
+        invoices_to_fetch = self.filtered(lambda inv: inv.l10n_ro_edi_state == 'invoice_sending')
+        self.env['res.company']._with_locked_records(invoices_to_fetch)
+
+        for invoice in invoices_to_fetch:
+            if errors := invoice._l10n_ro_edi_get_pre_send_errors():
+                to_delete_documents |= invoice._l10n_ro_edi_get_failed_documents()
+                invoice._l10n_ro_edi_create_document_invoice_sending_failed('\n'.join(errors))
+                continue
+
+            active_sending_document = invoice.l10n_ro_edi_document_ids.filtered(lambda d: d.state == 'invoice_sending')[0]
+            previous_raw = active_sending_document.attachment_id.sudo().raw
+            result = self.env['l10n_ro_edi.document']._request_ciusro_fetch_status(
+                company=invoice.company_id,
+                key_loading=active_sending_document.key_loading,
+                session=session,
+            )
+
+            if result == {}:
+                continue
+            elif 'error' in result:
+                to_delete_documents |= invoice._l10n_ro_edi_get_sending_and_failed_documents()
+                invoice._l10n_ro_edi_create_document_invoice_sending_failed(result['error'], previous_raw)
+            else:  # result == {'key_download': <str>}
+                # use the obtained key_download to immediately make a download request and process them
+                final_result = self.env['l10n_ro_edi.document']._request_ciusro_download_answer(
+                    company=invoice.company_id,
+                    key_download=result['key_download'],
+                    session=session,
+                )
+                to_delete_documents |= invoice._l10n_ro_edi_get_sending_and_failed_documents()
+                if 'error' in final_result:
+                    invoice._l10n_ro_edi_create_document_invoice_sending_failed(final_result['error'], previous_raw)
+                else:
+                    invoice._l10n_ro_edi_create_document_invoice_sent(final_result)
+
+            if not tools.config['test_enable'] and not modules.module.current_test:
+                self._cr.commit()
```

Comment:
I think you will lost the lock for all invoices when you commit so I think, you need to move the savepoint into the loop.
Idk how bad is it to have a savepoint in a loop if we commit right after the web-service :thinking: